flatsurf/geometry/gl2r_orbit_closure.py

```diff
@@ -792,7 +792,7 @@ def cylinder_circumference(self, component, A, sc_index, proj):
             sage: O.cylinder_circumference(c1, *kz) # optional: pyflatsurf
             (0, 0, -1, 0)
         """
-        if component.cylinder() != True:
+        if component.cylinder() is not True:
```

Comment:
We need to mark this with a `# noqa` comment or something like that. `component.cylinder()` is a ternary yes/no/maybe from boost, so the original `!=` is correct here.

Reply:
(and we should have a test that fails with the `is not`.)

Reply:
comparison to True using `==`or `!=` is never correct. If the variable v is boolean, then just use `if v`. If the variable can take other values, then the correct way is to use `is True` or `is not True`, that will check if the result is equal to the boolean constant or not

Reply:
I agree with you in general. But this is a somewhat mysterious type coming out of automatic translation into C++ code. The Python model does not apply here.

So, the variable is certainly not boolean. However, it does implement an `==` operator. And that operator knows how to compare to booleans.

Reply:
Let's write `x=component.cylinder()`.

So, `x is not True` is always true. Because it's not a boolean.

For the possible values of `x` we get the following table in Python:

| x | x | not x | == True | != True |
| -- | -- | -- | -- | -- |
| true | true | false | true | false |
| false | false | true | false | true |
| indeterminate | false | true | false | true |

So, `x != True` is the same as `not x`. But that's just because Python implements `not x` as `not bool(x)` afaik.

The corresponding table in C++ (where this type lives) is:

| x | x | !x | == true | != true |
| -- | -- | -- | -- | -- |
| true | true | false | true | false |
| false | false | true | false | true |
| indeterminate | false | **false** | false | true |

It's a semantic detail. And arguably, we should try to find a better solution here. But I think I want to keep a condition that also works in C++.

Reply:
That's the type in C++: https://www.boost.org/doc/libs/1_59_0/doc/html/tribool/tutorial.html